ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/ColumnReferenceValidator.java

```diff
@@ -85,46 +83,55 @@ public Void visitQualifiedColumnReference(
         final QualifiedColumnReferenceExp node,
         final Object context
     ) {
-      getSource(node.getLocation(), Optional.of(node.getQualifier()), node.getColumnName())
-          .ifPresent(referencedSources::add);
+      validateColumn(node);
       return null;
     }
 
-    private Optional<SourceName> getSource(
-        final Optional<NodeLocation> location,
-        final Optional<SourceName> sourceName,
-        final ColumnName name
-    ) {
-      final Set<SourceName> sourcesWithField = sourceSchemas.sourcesWithField(sourceName, name);
+    private void validateColumn(final ColumnReferenceExp colRef) {
+      final Set<SourceName> sourcesWithField = sourceSchemas
+          .sourcesWithField(colRef.maybeQualifier(), colRef.getColumnName());
+
       if (sourcesWithField.isEmpty()) {
-        throw new KsqlException(errorPrefix(location) + "column '"
-            + sourceName.map(n -> n.text() + KsqlConstants.DOT + name.text())
-            .orElse(name.text())
-            + "' cannot be resolved.");
+        if (couldBeSyntheticJoinColumn(colRef)) {
+          // Validating this is handled in the logical model.
+          return;
+        }
+
+        throw new UnknownColumnException(clauseType, colRef);
       }
 
-      if (sourcesWithField.size() > 1) {
-        final String possibilities = sourcesWithField.stream()
-            .map(source -> new QualifiedColumnReferenceExp(source, name))
-            .map(Objects::toString)
-            .sorted()
-            .collect(Collectors.joining(", "));
+      final SourceName source = colRef.maybeQualifier()
+          .orElseGet(() -> {
+            // AstSanitizer should catches ambiguous columns
+            return Iterables.getOnlyElement(sourcesWithField);
+          });
+
+      referencedSources.add(source);
+    }
 
-        throw new KsqlException(errorPrefix(location) + "column '"
-            + name.text() + "' is ambiguous. "
-            + "Could be any of: " + possibilities);
+    private boolean couldBeSyntheticJoinColumn(final ColumnReferenceExp colRef) {
+      if (!possibleSyntheticColumns) {
+        // Some queries never have synthetic columns, e.g. pull or aggregations.
+        return false;
       }
 
-      return Optional.of(Iterables.getOnlyElement(sourcesWithField));
-    }
+      if (!sourceSchemas.isJoin()) {
+        // Synthetic join columns only occur in joins... duh!
+        return false;
+      }
 
-    private String errorPrefix(final Optional<NodeLocation> location) {
-      final String loc = location
-          .map(Objects::toString)
-          .map(text -> text + ": ")
-          .orElse("");
+      if (colRef instanceof QualifiedColumnReferenceExp) {
+        // Synthetic join columns can't be qualified, as they don't belong to any source
+        return false;
+      }
+
+      if (!clauseType.equals("SELECT")) {
```

Comment:
seems like if we're checking equality it should be made either an enum or a constant

Reply:
Normally I'd agree, but it's only short term code and there are tests to ensure it does what we need.